nobody_home.groovy

```diff
@@ -1,164 +1,424 @@
-  }
+// event handler when the sunset time is reached
+def sunsetHandler(evt)
+{
+    // we store the mode we should be in, IF someone's home
+    state.modeIfHome = newSunsetMode
+    state.currentSunMode = "sunSet"
 
-  else {
-    log.debug("Mode is the same, not evaluating")
-  }
+    // change mode if someone's home, otherwise set to away
+    changeSunMode(newSunsetMode)
 }
 
-def presence(evt) {
-  if(evt.value == "not present") {
-    log.debug("Checking if everyone is away")
-
-    if(everyoneIsAway()) {
-      log.info("Starting ${newAwayMode} sequence")
-      def delay = (awayThreshold != null && awayThreshold != "") ? awayThreshold * 60 : 10 * 60
-      runIn(delay, "setAway")
+def changeSunMode(newMode)
+{
+    // if everyone is away, we need to check and ensure the system is
+    // in away mode.
+    if (isEveryoneAway()) {
+        // this shouldn not happen normally as the mode should already
+        // be changed during presenceHandler, but in case this is not
+        // done, such as when app is initially installed while away,
+        // and system is not in away mode, then we toggle it to away
+        // at the sun rise/set event.
+        changeMode(newAwayMode, " because no one is present")
+    } else {
+        // someone is home, we update the mode depending on
+        // sunrise/sunset.
+        if (state.currentSunMode == "sunRise") {
+            changeMode(newMode, " because it's sunrise")
+        } else if (state.currentSunMode == "sunSet") {
+            changeMode(newMode, " because it's sunset")
+        } else {
+            changeMode(newMode)
+        }
     }
-  }
+}
 
-  else {
-    if(location.mode != state.sunMode) {
-      log.debug("Checking if anyone is home")
+// ********** presence handling **********
 
-      if(anyoneIsHome()) {
-        log.info("Starting ${state.sunMode} sequence")
+// event handler when presence sensor changes state
+def presenceHandler(evt)
+{
+    // get the device name that resulted in the change
+    state.eventDevice= evt.device?.displayName
 
-        changeSunMode(state.sunMode)
-      }
+    // is setInitialMode() still pending?
+    if (state.pendingOp == "init") {
+        log.debug("pending ${state.pendingOp} op still in progress, ignoring presence event")
+        return
     }
 
-    else {
-      log.debug("Mode is the same, not evaluating")
+    if (evt.value == "not present") {
+        handleDeparture()
+    } else {
+        handleArrival()
     }
-  }
 }
 
-def setAway() {
-  if(everyoneIsAway()) {
-    if(location.mode != newAwayMode) {
-      def message = "${app.label} changed your mode to '${newAwayMode}' because everyone left home"
-      log.info(message)
-      send(message)
-      setLocationMode(newAwayMode)
+def handleDeparture()
+{
+    log.info("${state.eventDevice} left ${location.name}")
+
+    // do nothing if someone's still home
+    if (!isEveryoneAway()) {
+        log.info("Someone is still present, no actions needed")
+        return
```

Comment:
I can understood. At my work there's a preference for not having too many nested if/else blocks, especially for large functions and when we are checking many conditions in the beginning. In those cases the interesting action gets indented too much to the right if we use "if bad condition else ...". So we tend to use simple "if bad then return" style. The code is flatter and easier to maintain.
As you can see I've gotten use to writing in that style.. ;)